src/checkers/inference/solver/SolverEngine.java

```diff
@@ -54,266 +37,95 @@
-            final long graphBuildingStart = System.currentTimeMillis();
-            constraintGraph = generateGraph(slots, constraints, processingEnvironment);
-            final long graphBuildingEnd = System.currentTimeMillis();
-            StatisticRecorder.record(StatisticKey.GRAPH_GENERATION_TIME, (graphBuildingEnd - graphBuildingStart));
-            solution = graphSolve(constraintGraph, configuration, slots, constraints, qualHierarchy,
-                    processingEnvironment, formatTranslator);
-        } else {
-            underlyingSolver = createSolverAdapter(solverType, configuration, slots, constraints,
-                    processingEnvironment, lattice, formatTranslator);
-            solution = solve();
-        }
-
-        if (solution == null) {
-            // Solution should never be null.
-            ErrorReporter.errorAbort("Null solution detected!");
-        }
-
-        if (collectStatistic) {
-            Map<String, Integer> modelRecord = recordSlotConstraintSize(slots, constraints);
-            PrintUtils.printStatistic(StatisticRecorder.getStatistic(), modelRecord, solverType,
-                    useGraph, solveInParallel);
-            PrintUtils.writeStatistic(StatisticRecorder.getStatistic(), modelRecord, solverType,
-                    useGraph, solveInParallel);
-        }
-        return solution;
+    public enum SolverEngineArg implements SolverArg {
+        solvingStrategy,
+        solver,
+        collectStatistic;
     }
 
-    /**
-     * This method configures following arguments: backEndType, useGraph,
-     * solveInParallel, and collectStatistic
-     * 
-     * @param configuration
-     */
-    private void configureSolverArgs(final Map<String, String> configuration) {
+    private final String BACKEND_PACKAGE_PATH = SolverFactory.class.getPackage().getName();
+    private final String STRATEGY_PACKAGE_NAME = SolvingStrategy.class.getPackage().getName();
 
-        final String solverName = configuration.get(SolverArg.solver.name());
-        final String useGraph = configuration.get(SolverArg.useGraph.name());
-        final String solveInParallel = configuration.get(SolverArg.solveInParallel.name());
-        final String collectStatistic = configuration.get(SolverArg.collectStatistic.name());
+    protected SolverFactory createSolverFactory() {
+        final String solverPackageName = BACKEND_PACKAGE_PATH + "." + solverName.toLowerCase();
+        final String solverFactoryClassName = solverName + "SolverFactory";
 
-        solverType = solverName == null ? SolverType.MAXSAT : SolverType.getSolverType(solverName);
-        if (solverType == null) {
-            ErrorReporter.errorAbort("Integration of solver \"" + solverName + "\" has not been implemented yet.");
+        try {
+            Class<?> SolverFactoryClass = Class.forName(solverPackageName + "." + solverFactoryClassName);
+            return (SolverFactory) SolverFactoryClass.getConstructor().newInstance();
```

Comment:
Add javadoc in `SolverFactory`, for underlying sovlerFactory should have a zero param constructor for reflectively loading here.

Reply:
Done.